File: services/Chat Orchestrator/src/domain.py
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone
from enum import Enum
from uuid import uuid4


class Role(str, Enum):
	anonymous_user = "anonymous_user"
	registered_user = "registered_user"
	operator = "operator"
	specialist = "specialist"
	system = "system"


class ChatStatus(str, Enum):
	open = "open"
	waiting_operator = "waiting_operator"
	in_progress_operator = "in_progress_operator"
	specialist_review = "specialist_review"
	resolved = "resolved"
	blocked = "blocked"
	closed = "closed"


class DeliveryTarget(str, Enum):
	rag_engine = "rag_engine"
	operator_queue = "operator_queue"
	specialist_queue = "specialist_queue"
	direct_user_reply = "direct_user_reply"
	chat_state_change = "chat_state_change"


class OperatorAction(str, Enum):
	close_chat = "close_chat"
	block_chat = "block_chat"
	resolve_chat = "resolve_chat"
	send_to_specialist_queue = "send_to_specialist_queue"


class SpecialistDecision(str, Enum):
	approve = "approve"
	reject = "reject"


@dataclass(slots=True)
class AccessDecisionEntity:
	allowed: bool
	reason: str


@dataclass(slots=True)
class RAGResultEntity:
	chunk_id: int | str
	document_id: int | str
	document_title: str
	chunk_index: int
	score: float
	text: str


@dataclass(slots=True)
class MessageEntity:
	chat_id: str
	sender_role: Role
	sender_id: str
	text: str
	recipient_role: Role
	message_id: str = field(default_factory=lambda: str(uuid4()))
	created_at: datetime = field(default_factory=lambda: datetime.now(timezone.utc))


@dataclass(slots=True)
class UserMessageResultEntity:
	chat_id: str
	route: DeliveryTarget
	chat_status: ChatStatus
	message: str
	rag_results: list[RAGResultEntity]
	queue_item_id: str | None = None


@dataclass(slots=True)
class ActionResultEntity:
	chat_id: str
	chat_status: ChatStatus
	message: str
	queue_item_id: str | None = None


@dataclass(slots=True)
class SpecialistReviewResultEntity:
	queue_item_id: str
	decision: SpecialistDecision
	knowledge_base_update_requested: bool
	message: str

File: services/Web Service/src/domain.ts
export type UserRole = "registered_user" | "operator" | "admin";

export type DeliveryRoute = "rag_engine" | "operator_queue";

export type ChatStatus =
  | "open"
  | "waiting_operator"
  | "in_progress_operator"
  | "closed"
  | "blocked"
  | "resolved"
  | "specialist_review";

export type OperatorAction = "close_chat" | "block_chat" | "resolve_chat" | "send_to_specialist_queue";

export interface ChatMessage {
  messageId: string;
  chatId: string;
  senderRole: UserRole | "assistant" | "system";
  senderId: string;
  text: string;
  createdAt: string;
}

export interface ChatThread {
  chatId: string;
  title: string;
  ownerUserId: string;
  status: ChatStatus;
  escalatedToOperator: boolean;
  assignedOperatorId: string | null;
  messages: ChatMessage[];
  createdAt: string;
  updatedAt: string;
}

export interface ChatSummary {
  chatId: string;
  title: string;
  status: ChatStatus;
  escalatedToOperator: boolean;
  assignedOperatorId: string | null;
  updatedAt: string;
  preview: string;
  userMessageCount: number;
}

export type KnowledgeRequestStatus = "pending" | "approved" | "rejected";
export type KnowledgeDispatchStatus = "queued" | "sent";

export interface KnowledgeRequest {
  requestId: string;
  chatId: string;
  question: string;
  answer: string;
  createdBy: string;
  createdAt: string;
  status: KnowledgeRequestStatus;
  reviewedBy: string | null;
  reviewedAt: string | null;
  dispatchStatus: KnowledgeDispatchStatus;
}

export interface AccountRecord {
  userId: string;
  login: string;
  role: UserRole;
  isBlocked: boolean;
  updatedAt: string;
}
